src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+use std::cmp::{Ordering, min, max};
+use std::f64::INFINITY;
+use std::mem::replace;
+
+use {Point};
+
+
+/// A function to compute the inter-cluster distance between set `A` and `B`.
+pub type LinkageFunction = Fn(&Fn(usize, usize) -> f64, &[usize], &[usize]) -> f64;
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+#[derive(Copy, Clone)]
+pub enum LinkageCriterion<'a> {
```

Comment:
I'm not totally sure if this sacrifices expressiveness in a major way, but I haven't played around with it, so I can't be sure.

Despite the comment about the benchmark, I think I'd be happy to just land as is, and then maybe tweak in future.